### ISSUE TITLE
[Discussion] Introduce some way to use module constructors through C#

### ISSUE THREAD

**Issue by john-h-k** [dotnet/csharplang#2486]
IL supports the notion of a 'module constructor', similar to a static constructor, but it is free (no attached class), and named the same (`specialname rtspecialname .cctor`). It is guaranteed to run before any code or type initialization. It could be exposed through C# via an attribute on a method or a special name (`DllMain`?). I reasoned this is more suitable for here rather than dotnet/Rosyln as it would be a language feature like static ctors or PInvoke even if exposed through an attribute.

* Basically no downsides I can see
* Useful in some situations

**Comment by svick** [comment]
> Basically no downsides I can see

That's not sufficient for a feature to be considered.

> Useful in some situations

What are those?

**Comment by marksmeltzer** [comment]
As a framework developer, this is an amazing proposal. I have been forced to come up with several convoluted workarounds to get the utility of static read-only fields without the substantial runtime penalties they come with.

Currently the runtime takes an init lock which is used to double check whether the static construction logic has been processed. Adding even one static readonly field to a class instantly adds that overhead to every external use of any member of that class. 

Having the ability to run initialization logic in a guaranteed, predictable order that has zero post module init runtime performance would be a huge benefit.

The Midori project that Duffy and team worked on had a similar performance benefit for static read-only constructions - they were even shared across the memory space of disparate processes since they were provably identical, which had additional performance benefits. 

Module initializers seem like a fantastic compromise: a module could be initializated in context (have access to prior application state that is already initializated - for example, environmental variables, etc.) and thus separate app domains or processes running the same init code could produce different results. That's useful. 

Ideally any readonly static fields that are accessed during this module init phase would be flagged and recognized by the JIT.

After module init is completed, the runtime init checks for all static access to the flagged classes can safely be omitted.

During init, the normal rules, including init checks, would apply to static field access.

This seems to have litte relevance to C# though. I imagine an assembly attribute would be added or perhaps a module "main" function. If we went the module main function, there could be some C# convention established to identify this unique method signature so that the compiler could produce an error if more than one such method were declared within a single module (which generally means per assembly, but principle not always since a single assembly can technically include many modules).

I don't have a strong opinion on whether to make this a C# feature. However, the special flagging I described above for static field access for JIT optimization is definitely a Roslyn feature opportunity.

**Comment by HaloFour** [comment]
@marksmeltzer 

Module initializers already exist in .NET and they have a lazy-init behavior like static type constructors do.  That could probably be an argument unto itself but I gather this language proposal is about exposing the existing framework feature.

**Comment by marksmeltzer** [comment]
Again, I don't see this as an obvious C# proposal, but if this were made accessible to C# developers it would have a huge benefit.

Moreover, the CoreFX team could add module init methods to enable key, high traffic static real-only fields to be preinitialized, which would have significant downstream cost savings for all applications.

That brings up the possibility for more pay to play enhancement opportunities where like classes could be moved to function specific modules. That would avoid forcing developers that don't use a feature pay to initialize it. Moreover, relocating code between modules is already a fully supported scenario that doesn't break existing applications.

**Comment by HaloFour** [comment]
@marksmeltzer 

I think it'd have to be a C# proposal in that the compiler would be required to emit the global function with the correct name and metadata for the framework to recognize and execute on assembly load.  C# doesn't permit writing global functions and C# doesn't permit functions with the required name.  So even if it's accomplished via pseudo-attributes and requires no new syntax, it does require specification.

**Comment by marksmeltzer** [comment]
@halofour

This makes me wonder about what else I don't know! 

Regardless of this proposal, I am going to immediately begin work to see of I can create a Roslyn extension to create an attribute that can be applied to a static method and have it compiled as a module init.

I am guessing that my suggestion above to flag static field access during init to eliminate post module initialization checks by the JIT during runtime would be a something the CoreCLR/JIT team would have to build though.

Since CoreCLR just recently got progressive JIT, I presume this scenario is not yet implemented. This feature would require (I think) progressive JIT, or at least something similar, in order to achieve it in the full framework.

**Comment by marksmeltzer** [comment]
If I cannot figure a way to do module init using a simple attribute with Roslyn extensions, I am positive that I can offer the desired feature via a nuget enabled post compile step using Cecil. (That's more heavy handed, but I will be shocked if I cannot achieve it using that approach.)

Thanks @johnkellyoxford for the inspiration!

**Comment by svick** [comment]
@marksmeltzer [Or you can use code that someone else already wrote.](https://github.com/kzu/InjectModuleInitializer)

**Comment by marksmeltzer** [comment]
@svick yes!!

File that under #themoreidontknow

I had planned to search for such a thing beflre attempting it 🤣

**Comment by marksmeltzer** [comment]
@svick I am still going to see about achieving it through a Roslyn extension attribute. The solution above uses Cecil, and seems not to have been updated in a while. Some issues raised on the project report problems with the debugging experience - probably due to the IL rewriting from Cecil. Will first have to check to see if it is up to date with current Cecil work, etc. It the current solution works as is or with slight modification that would be nice.

Having a Roslyn based solution would be nicer though.

**Comment by johnkellyoxford** [comment]
It is a C# proposal because the compiler would need to emit the correct signature and importantly, do it outside of a class, something that is currently un doable from normal C#

**Comment by pinkfloydx33** [comment]
We are currently using Fody to inject a module initializer in a few projects. It works great, but to someone unfamiliar with what's going on the set up looks a little weird. We also need to make sure nobody changes the method name (ton of comments around the method) otherwise the build breaks. It'd be nice to have as a first class language feature, but in the meantime we'll survive using Fody (especially since we use it for other things)

**Comment by masonwheeler** [comment]
Having module initializers run lazily is a poor implementation that ought to be fixed.

Class constructors run lazily because C# (and the CLR in general) does not have a mechanism to specify a well-defined initialization order, so lazy init is the simplest way to avoid order-of-operation problems.  With module/assembly initialization, this is not an issue, as the loader gives us a well-defined order and ensures that all dependency assemblies will be loaded before the assemblies they depend on.  Therefore, lazy loading is not necessary for module initializers.

And not only is it unnecessary, it's actively detrimental as it requires external code to reach into the assembly and touch something inside of it before it will run.  This makes things like plugin development unnecessarily complex: instead of a plugin-holding assembly simply using a module constructor to register the plugins it provides with the plugin system at load time, the plugin system needs to use some sort of Reflection-based technique to scan all assemblies being loaded in order to see if any of them contain plugins.

**Comment by CyrusNajmabadi** [comment]
> This makes things like plugin development unnecessarily complex: instead of a plugin-holding assembly simply using a module constructor to register the plugins it provides with the plugin system at load time, the plugin system needs to use some sort of Reflection-based technique to scan all assemblies being loaded in order to see if any of them contain plugins.

That doesn't seem unnecessarily complex?  That sounds pretty simple and well understood (and it's something with ample solutions already present that can be used).  For example, Roslyn itself runs off of the [MEF](https://docs.microsoft.com/en-us/dotnet/framework/mef/) system for this.  It's simple, flexible, and does a great job at allowing plugins with minimal fuss or effort.  

It seems strange to add a new feature to support a use case that already has good solutions for it.

**Comment by masonwheeler** [comment]
Reflection is not a good solution; it's a hack that exists to fill in the gaps where good solutions do not exist.

Being able to do this without needing reflection would be a good solution.

**Comment by HaloFour** [comment]
@masonwheeler 

> Reflection is not a good solution; it's a hack that exists to fill in the gaps where good solutions do not exist.

Reflection is a perfectly good solution.  You not liking it does not make it a hack.

**Comment by johnkellyoxford** [comment]
@masonwheeler what? They are run lazily but still following the strict rules about when they can be run. You can just `RuntimeHelpers.RunModuleConstructor` if you need to

**Comment by CyrusNajmabadi** [comment]
Something that is flexible, and works well for a wide variety of situations doesn't feel like a hack.  it feels like a hack to me to have to have a dedicated solution for such a small problem space when a more general solution exists and is easy to pick up and use.

**Comment by masonwheeler** [comment]
@johnkellyoxford Sure, but that should be unnecessary given that the reason why they need to run lazily does not exist in this case.

**Comment by masonwheeler** [comment]
@CyrusNajmabadi 
> Something that is flexible, and works well for a wide variety of situations doesn't feel like a hack. 

Depends.  Reflection is a general-purpose mechanism that can accomplish essentially *any* purpose in general, but only by doing so much more slowly and code-intensively than code that specifically accomplishes that same purpose.

In other words, it's an inefficient way to hack around the lack of specific-purpose code, and therefore specific-purpose code should be preferred whenever possible.

**Comment by CyrusNajmabadi** [comment]
> Depends. Reflection is a general-purpose mechanism that can accomplish essentially any purpose in general, but only by doing so much more slowly and code-intensively than code that specifically accomplishes that same purpose.

Seems like an acceptable place to land.  Can something like Fody be used for these purposes?  It seems like it could just take some properly attributed static method, and transform it into the 'module constructor'.

**Comment by CyrusNajmabadi** [comment]
Wait, it looks like it's already in Fody and very simple to use: https://github.com/Fody/ModuleInit

Why does the language need anything here?

**Comment by pinkfloydx33** [comment]
@CyrusNajmabadi see my comment above. That's what we are currently using in some of our projects. We have no issues with it other than the fact it needs to be well documented for those unfamiliar with Fody/module initializers (no big deal). Though for folks that aren't using Fody already, they need to introduce it into their build. Like I said a solution already exists, but it'd be "nice" if this was a first class feature.

**Comment by CyrusNajmabadi** [comment]
> and therefore specific-purpose code should be preferred whenever possible

That is not true at all.  A general maxim of our language development is that features need to prove their worth.  That means they should provide value that can either be broadly applicable (i.e. you'd expect tons of the user base to use it), or provide such exceptionally high value, that's it's ok for it to be for a narrower slice of users.  Examples of the former are things like linq, lambdas, patterns, slices, etc.  It makes the programming model much nicer, and is used by a huge amount of the ecosystem.  Examples of the latter are things like all the recent `ref/span/readonly` work.  It's not expected to be used by that many, but for the audience it is for, it's staggeringly beneficial.  And those particular areas are high impact (games, web servers, etc.).

This is absent here.  The number that would benefit is quite small, and the value on top of what's already possible is tiny.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi see my comment above. That's what we are currently using in some of our projects. We have no issues with it other than the fact it needs to be well documented for those unfamiliar with Fody/module initializers (no big deal). 

Seems acceptable :)   They'd likely need to be well-documented even if they were in the language as well :D

> Though for folks that aren't using Fody already, they need to introduce it into their build. 

Seems acceptable to me.  

 >Like I said a solution already exists, but it'd be "nice" if this was a first class feature.

Sure.  However, see above.  It's especially relevant to take the -100 rule into account: https://blogs.msdn.microsoft.com/ericgu/2004/01/12/minus-100-points/

Effectively, language features are enormously costly.  It's really vital that they be *valuable* enough to justify that cost.   See https://github.com/dotnet/csharplang/issues/2486#issuecomment-489470580 for why i think that value isn't present *esp* given there are preexisting solutions people can use.

I def know that everyone wants their pet feature to be "in the language".  but it simply can't be designed that way.  That would add hundreds to thousands of bits that would make the language nearly impossible to move forward with in a reasonable fashion.  So the language is *extremely* selective about what gets added.  Things make it only if tehre is exceptionally strong belief that this will be hugely valuable.

**Comment by masonwheeler** [comment]
> The number that would benefit is quite small

It's hard to say that with any certainty.  Right now, very few people have expressed any desire to use it because most people don't know about it, because it isn't available in the C# language.  If it was, people would find more use for it.

Remember how, back in the .NET 2.0 days, [the team considered Generics to be "for academics only" with no practical use, while the "killer feature" for the new CLR version was SQLCLR?](https://blogs.msdn.microsoft.com/dsyme/2011/03/15/netc-generics-history-some-photos-from-feb-1999/)  Things turned out very differently once it became available to real users.

**Comment by CyrusNajmabadi** [comment]
> It's hard to say that with any certainty. Right now, very few people have expressed any desire to use it because most people don't know about it, because it isn't available in the C# language. If it was, people would find more use for it.

it's been 20 years, the language team has large sources of inputs.  People are not shy about proposing features or indicating their needs. 

> Things turned out very differently once it became available to real users.

Generics were felt to prove their worth prior to being done.  That's why there were done in the first place.

**Comment by CyrusNajmabadi** [comment]
> It's hard to say that with any certainty. Right now, very few people have expressed any desire to use it because most people don't know about it, because it isn't available in the C# language. If it was, people would find more use for it.

So how would you decide which language features to do?  If they aren't in the language, and only have a couple of people asking for them... you'd just do it?  

How do you decide which features are worth it or not?

**Comment by CyrusNajmabadi** [comment]
For generics, btw, those proposing it could should wide swaths of applicable scenarios that would benefit, and could show how much benefit a particular user would be expected to get.  It fit into both categories above in that it was both broadly applicable to a huge amount of the user base *and* it was expected to have *high* impact across the board for that user base.  So it was a double-whammy of a feature.

That's why it was done *even though* it was the only time (up until now) that there was a break in the CLR.  That's *just* how valuable it was.

**Comment by CyrusNajmabadi** [comment]
Note: i want to reiterate the very first response to this request from @svick:

> > Basically no downsides I can see

> That's not sufficient for a feature to be considered.

Agreed.  See the -100 approach to language design

> >  Useful in some situations

> What are those?

I would like this answered as well.  So far, i've heard about plugins.  And i've heard that this is already possible ith fody... so i'm not seeing the usefulness.  It would be good to have some very strong arguments about why this is needed, and what can demonstrate that it really is valuable enough to be added to the language.  Thanks!

**Comment by yaakov-h** [comment]
@CyrusNajmabadi 

> A general maxim of our language development is that features need to prove their worth. That means they should provide value that can either be broadly applicable ..., or provide such exceptionally high value

I dunno, this kind of feature seems to have the same amount of value as `private protected` did, and in theory one could have achieved `private protected` with a post-compilation rewriter like Fody, too.

This would ideally be a small-ish feature that allows you to express an IL construct directly in C#. The major difference would be that Module Constructors aren't currently a pain point for the Roslyn team, but `private protected` was.

**Comment by CyrusNajmabadi** [comment]
> The major difference would be that Module Constructors aren't currently a pain point for the Roslyn team, but private protected was.

And `private-protected` (né protected-and-internal) was something known to be valuable and wanted by a much larger group of people for a very long time.  Indeed, the LDM was literally on board with this for probably upwards of a decade.  The only thing holding it back was literally just a design.  i.e. no one really liked the idea of the syntax being any of:

1. `proternal`
2. `protected_and_internal`
3. `protected and internal`
4. `protectedandinternal`

etc. etc.

there was consistent demand for this over the years, but we were just blocked on "how do we actually put it in the language?  in teh end, we couldn't find a way that didn't make someone retch.  But we accepted that that was ok, and the pros of having it outweighed the cons of `private protected` not actually making any sense.

**Comment by marksmeltzer** [comment]
@CyrusNajmabadi

I specified some examples in this thread of how framework authors can use module init that had nothing to do with plug-in support.

**Comment by marksmeltzer** [comment]
@CyrusNajmabadi as I said in my long comment, there would be an opportunity to do an Midori style optimization to remove the static init lock checks from all static classes initialized by the module constructor - that would have a very large and quantifiable perf benefit for frequently accessed classes.

Making those scenarios play to pay would involve moving relevant code to narrower modules as needed using forwarding. 

Now that we have progressive JIT, such a thing would be possible. Making the module constructor a first class citizen would be nice while we're at it - but I agree that it isn't strictly necessary since it would only be used by a small percentage of users. 

However, it strikes me as something that could help motivate the performance improvement opportunity if it were a first class citizen.

**Comment by marksmeltzer** [comment]
BTW, I posted in the CLR project a proposal to use progressive JIT to eliminate the locks surrounding static access -- https://github.com/dotnet/coreclr/issues/24571#issue-444061096

That concept is generalizable beyond just module constructors.

**Comment by gafter** [comment]
Feature championed in https://github.com/dotnet/csharplang/issues/2608

**Comment by YairHalberstadt** [comment]
Closing as championed in https://github.com/dotnet/csharplang/issues/2608
